modal/sandbox.py

```diff
@@ -48,47 +65,72 @@ async def read(self) -> str:
+
+        return data
+
+    async def _get_logs(self) -> AsyncIterator[Optional[api_pb2.TaskLogs]]:
+        """mdmd:hidden
+        Streams sandbox logs from the server to the reader.
+
+        When the stream receives an EOF, it yields None. Once an EOF is received,
+        subsequent invocations will not yield logs.
+        """
+        if self.eof:
+            yield None
+            return
 
-        async def _get_logs():
-            nonlocal last_log_batch_entry_id, completed, data
+        completed = False
 
+        retries_remaining = 10
+        while not completed:
             req = api_pb2.SandboxGetLogsRequest(
                 sandbox_id=self._sandbox_id,
                 file_descriptor=self._file_descriptor,
                 timeout=55,
-                last_entry_id=last_log_batch_entry_id,
+                last_entry_id=self._last_log_batch_entry_id,
             )
-            log_batch: api_pb2.TaskLogsBatch
-            async for log_batch in unary_stream(self._client.stub.SandboxGetLogs, req):
-                if log_batch.entry_id:
-                    # log_batch entry_id is empty for fd="server" messages from AppGetLogs
-                    last_log_batch_entry_id = log_batch.entry_id
-
-                if log_batch.eof:
-                    completed = True
-                    break
-
-                for item in log_batch.items:
-                    data += item.data
-
-        while not completed:
             try:
-                await _get_logs()
+                async for log_batch in unary_stream(self._client.stub.SandboxGetLogs, req):
+                    self._last_log_batch_entry_id = log_batch.entry_id
+
+                    for message in log_batch.items:
+                        yield message
+                    if log_batch.eof:
+                        self.eof = True
+                        completed = True
+                        yield None
+                        break
             except (GRPCError, StreamTerminatedError) as exc:
-                if isinstance(exc, GRPCError):
-                    if exc.status in RETRYABLE_GRPC_STATUS_CODES:
+                if retries_remaining > 0:
+                    retries_remaining -= 1
+                    if isinstance(exc, GRPCError):
+                        if exc.status in RETRYABLE_GRPC_STATUS_CODES:
+                            await asyncio.sleep(1.0)
+                            continue
+                    elif isinstance(exc, StreamTerminatedError):
                         continue
-                elif isinstance(exc, StreamTerminatedError):
-                    continue
                 raise
 
-        return data
+    def __aiter__(self):
+        """mdmd:hidden"""
+        self._stream = self._get_logs()
+        return self
+
+    async def __anext__(self):
+        """mdmd:hidden"""
+        value = await self._stream.__anext__()
+
+        # The stream yields None if it receives an eof batch.
```

Comment:
```suggestion
        # The stream yields None if it receives an EOF batch.
```

nit: consistency throughout module